man/capmarker.1
.TH capmarker 1 "01 Feb 2012" "BTH" "Measurement Area Manual"
.SH NAME
capmarker \- Send a special packet that be identified in the capture stream.
.SH SYNOPSIS
.nf
.B capmarker \-e \fIEXP\fP \-r \fIRUN\fP [\-k \fIKEY\fP] [\-s \fISEQ\fP]
.SH DESCRIPTION
Markers are special packets that \fIlibcap_utils\fR can easily
identify and which contains meta-data about the traffic. \fIcapdump\fR
can use markers to split captures to multiple files and it can be used
to later identify what experiment and run the traffic belongs to.

.nf
\fB\-e\fR, \fB\-\-experiment\fR=\fIID\fR
\fB\-r\fR, \fB\-\-run\fR=\fIID\fR
\fB\-k\fR, \fB\-\-key\fR=\fIID\fR
\fB\-s\fR, \fB\-\-sequence\fR=\fIID\fR
.RS
Extra information passed in the marker. There is no inherit meaning to
these numbers, it is considered domain-specific information that the
user puts meaning into.
.BR
Usually all numbers starts at 1 and for each new scenario the
\fBexperiment\fR number is increased. If the scenario consists of
multiple runs the \fBrun\fR number is reset back to 1 at the start and
increased once per run. \fBKey\fR can be used as an identifier and
\fBsequence\fR might be a global sequential number.
.RE
.TP
\fB\-c\fR, \fB\-\-comment\fR
Include text-comment in marker. Up to 63 characters can be included verbatim.
.TP
\fB\-u\fR
Send marker as a UDP packet. Default.
.TP
\fB\-t\fR
Send marker as a TCP packet (need a receiver for to work, e.g. `nc -lp PORT`).
.TP
\fB\-x\fR
Send marker as an ethernet frame.
.TP
\fB\-T\fR, \fB\-\-terminate\fR
Include a hint to consumers that this stream is no longer interesting,
e.g. if you have some automatic tests or scenario that have finished
you make include the termination hint so consumers can stop
processing.
.TP
\fB\-h\fR, \fB\-\-help\fR
Short help.
.SH COPYRIGHT
Copyright (C) 2011-2012 David Sveningsson <dsv@bth.se>.
.SH "SEE ALSO"
stream-address(3), capplay(1)


NEWS
caputils-0.7.15
---------------

  * pf_ring: compatibility with 6.0

caputils-0.7.14
---------------

  * adding a custom ethernet type for DPMI diagnostic packets.
  * capshow supports many additional protocols.
  * general improvements to output from tools.
  * capdump improved filename generation when using markers.
  * version number always includes git revision if available.
  * new local-only filters for DT and frame numbers.
  * adding generic functions for printing, stateful.
  * pfring always using MTU buffer size instead of hardcoded value.
  * fix issues when doing capture on loopback interfaces.
  * general build fixes.

caputils-0.7.13
---------------

  * pcap2cap improved.
  * Support for BPF filters.
	* capinfo showing more data.

caputils-0.7.12
---------------

  * capfilter silent output.
  * capmerge regression fixes.

caputils-0.7.11
---------------

  * capfilter fixes.

caputils-0.7.10
---------------

  * Additional capshow package decoding.
  * Added stream_peek, stream_num_addresses, timepico_now.
  * stream_copy truncates packages to caplen.

caputils-0.7.9
--------------

  * capinfo output improvements.
  * capdump output improvements.
  * Drops invalid measurement frames.

caputils-0.7.8
--------------

  * Quick bugfixes for PF_RING.

caputils-0.7.7
--------------

  * More statistics and debugging messages.
  * Support for AND/OR filters.
  * Additional functions for developers
  * support for PF_RING when doing ethernet capture.

caputils-0.7.6
--------------

  * caphead_t typedef
  * prototype for stream_from_getopt changed, abi breakage.
  * additional functions for dealing with picotime
  * capshow and capfilter improvements.

caputils-0.7.5
--------------

  * `stream_t` is made a pointer, this will break applications but is easily
    fixed.
  * Additional utility functions.
  * Changed buffering strategy.
  * Added capmarker and capshow.

caputils-0.7.4
--------------

  * Integrating libmarc into caputils.
  * Adding function to add multiple ethernet multicast memberships to a single stream
  * typedef struct stream stream_t

caputils-0.7.3
--------------

	* Moving {,v}logmsg from libmarc here.
	* Major API changes; breaks ABI.
	* Fixed capdump.
	* More man-pages.
	* Fixing compability issues with linux-2.4

caputils-0.7.2
--------------

	* Split filtering code from libcaputils to libfilter.
	* Added two more utilities:
	    - capdump - record a stream
	    - capfilter - filter existing stream
	* Added manpages.

caputils-0.7.0
--------------

Based on the 0.5 branch.
	* Backwards-compability support
	* Numerous API-fixes, the API is *NOT* compatable with earlier versions, but there
	  is only minor fixes required.
	* Numerous bug-fixes, including endianess-fixes.
	* New tools: capinfo, shows information about capfiles.
	* Uses autotools and pkg-config


tools/capmarker.c
/**
 * libcap_utils - DPMI capture utilities
 * Copyright (C) 2003-2013 (see AUTHORS)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

#include "caputils/marker.h"
#include "caputils/utils.h"
#include "caputils/version.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <getopt.h>
#include <errno.h>
#include <signal.h>
#include <unistd.h>
#include <sys/stat.h>
#include <arpa/inet.h>
#include <pwd.h>
#include "be64toh.h" /* for compability */

static enum {
	MODE_UDP,
	MODE_TCP,
	MODE_ETH,
} mode = MODE_UDP;

static struct marker marker = {
	.magic = MARKER_MAGIC,
	.version = 1,
	.flags = 0,
	.reserved = 0,
	.exp_id = 0,
	.run_id = 0,
	.key_id = 0,
	.seq_num = 0,
	.timestamp = 0,
};

static const char* program_name = NULL;
static const char* shortopts = "e:r:k:s:c:utxTh";
static struct option longopts[]= {
	{"experiment", required_argument, 0, 'e'},
	{"run",        required_argument, 0, 'r'},
	{"key",        required_argument, 0, 'l'},
	{"sequence",   required_argument, 0, 's'},
	{"comment",    required_argument, 0, 'c'},
	{"terminate",  no_argument,       0, 'T'},
	{"help",       no_argument,       0, 'h'},
	{0, 0, 0, 0} /* sentinel */
};

static void show_usage(void){
	printf("capmarker-%s\n", caputils_version(NULL));
	printf("(C) 2012 David Sveningsson <david.sveningsson@bth.se>\n");
	printf("Usage: %s [OPTIONS] IP:PORT..\n", program_name);
	printf("  -e, --experiment=ID  Current experiment ID.\n"
	       "  -r, --run=ID         Current run ID.\n"
	       "  -k, --key=INT        Domain information. [default: 0]\n"
	       "  -s, --sequence       Sequence start number. [default: 0].\n"
	       "  -c, --comment        Comment\n"
	       "  -u                   UDP packet [default]\n"
	       "  -t                   TCP packet\n"
	       "  -x                   Ethernet packet\n"
	       "  -T, --terminate      Set termination flag\n"
	       "  -h, --help           This text.\n");
}

static int send_udp(const struct in_addr dst, in_port_t port){
	/* open socket */
	int sd = socket(AF_INET, SOCK_DGRAM, 0);
	if ( sd == -1 ){
		fprintf(stderr, "%s: failed to open socket: %s\n", program_name, strerror(errno));
		return 1;
	}

	int on = 1;
	if (setsockopt(sd, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(int)) < 0) {
		fprintf(stderr, "%s: setsockopt(SO_REUSEADDR) failed: %s\n", program_name, strerror(errno));
		return 1;
	}

	/* setup source address */
	static struct sockaddr_in src_addr;
	memset(&src_addr, 0, sizeof(struct sockaddr_in));
	src_addr.sin_family = AF_INET;
	src_addr.sin_port = (in_port_t)htons(MARKERPORT);
	src_addr.sin_addr.s_addr = htonl(INADDR_ANY);

	/* setup destination address */
	static struct sockaddr_in dst_addr;
	memset(&dst_addr, 0, sizeof(struct sockaddr_in));
	dst_addr.sin_family = AF_INET;
	dst_addr.sin_port = port;
	dst_addr.sin_addr.s_addr = dst.s_addr;

	if ( bind(sd, &src_addr, sizeof(struct sockaddr_in)) == -1 ){
		fprintf(stderr, "%s: failed to bind socket: %s\n", program_name, strerror(errno));
		return 1;
	}

	/* send marker */
	if ( sendto(sd, &marker, sizeof(struct marker), 0, (struct sockaddr*)&dst_addr, sizeof(struct sockaddr_in)) == -1 ){
		fprintf(stderr, "%s: sendto failed: %s\n", program_name, strerror(errno));
	}

	return 0;
}

static int send_tcp(const struct in_addr dst, in_port_t port){
  /* open socket */
  int sd = socket(AF_INET, SOCK_STREAM, 0);
  if ( sd == -1 ){
    fprintf(stderr, "%s: failed to open socket: %s\n", program_name, strerror(errno));
    return 1;
  }

  int on = 1;
  if (setsockopt(sd, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(int)) < 0) {
    fprintf(stderr, "%s: setsockopt(SO_REUSEADDR) failed: %s\n", program_name, strerror(errno));
    return 1;
  }

  /* setup source address */
  static struct sockaddr_in src_addr;
  memset(&src_addr, 0, sizeof(struct sockaddr_in));
  src_addr.sin_family = AF_INET;
  src_addr.sin_port = (in_port_t)htons(MARKERPORT);
  src_addr.sin_addr.s_addr = htonl(INADDR_ANY);

  /* setup destination address */
  static struct sockaddr_in dst_addr;
  memset(&dst_addr, 0, sizeof(struct sockaddr_in));
  dst_addr.sin_family = AF_INET;
  dst_addr.sin_port = port;
  dst_addr.sin_addr.s_addr = dst.s_addr;

  if ( bind(sd, &src_addr, sizeof(struct sockaddr_in)) == -1 ){
    fprintf(stderr, "%s: failed to bind socket: %s\n", program_name, strerror(errno));
    return 1;
  }



  if ( connect(sd, &dst_addr, sizeof(struct sockaddr_in)) == -1 ){
    fprintf(stderr, "%s: failed to bind socket: %s\n", program_name, strerror(errno));
    return 1;
  }

  fprintf(stderr, "%s: Sending from %d to %d \n",program_name,MARKERPORT,port);

  /* send marker */
  if ( sendto(sd, &marker, sizeof(struct marker), 0, (struct sockaddr*)&dst_addr, sizeof(struct sockaddr_in)) == -1 ){
    fprintf(stderr, "%s: sendto failed: %s\n", program_name, strerror(errno));
  }

  close(sd);
  return 0;
}








int main(int argc, char **argv){
	/* extract program name from path. e.g. /path/to/MArCd -> MArCd */
	const char* separator = strrchr(argv[0], '/');
	if ( separator ){
		program_name = separator + 1;
	} else {
		program_name = argv[0];
	}

	/* reset marker */
	marker.timestamp = htobe64(time(NULL));
	marker.flags = 0;
	memset(marker.comment, 0, 64);

	int op, option_index = -1;
	while ( (op = getopt_long(argc, argv, shortopts, longopts, &option_index)) != -1 ){
		switch (op){
		case 0:   /* long opt */
		case '?': /* unknown opt */
			break;

		case 'e':
			marker.exp_id = atoi(optarg);
			if ( marker.exp_id == 0 ){
				fprintf(stderr, "%s: Experiment ID must be greater than 0.\n", program_name);
				return 1;
			}
			break;

		case 'r':
			marker.run_id = atoi(optarg);
			if ( marker.run_id == 0 ){
				fprintf(stderr, "%s: Run ID must be greater than 0.\n", program_name);
				return 1;
			}
			break;

		case 'k':
			marker.key_id = atoi(optarg);
			break;

		case 's':
			marker.seq_num = atoi(optarg);
			break;

		case 'c':
			strncpy(marker.comment, optarg, 63); /* 63 so that 64 will always be a NULL-terminator */
			break;

		case 'u':
			mode = MODE_UDP;
			break;

		case 't':
			mode = MODE_TCP;
			break;

		case 'x':
			mode = MODE_ETH;
			break;

		case 'T': /* --terminate */
			marker.flags |= MARKER_TERMINATE;
			break;

		case 'h':
			show_usage();
			exit(0);
			break;

		default:
			if ( option_index >= 0 ){
				fprintf(stderr, "flag --%s declared but not handled\n", longopts[option_index].name);
			} else {
				fprintf(stderr, "flag -%c declared but not handled\n", op);
			}
			abort();
		}
		option_index = -1;
	}

	/* network order */
	marker.magic = htonl(marker.magic);
	marker.reserved = htons(marker.reserved);
	marker.exp_id = htonl(marker.exp_id);
	marker.run_id = htonl(marker.run_id);
	marker.key_id = htonl(marker.key_id);
	marker.seq_num = htonl(marker.seq_num);

	/* ensure at least one destination exists */
	if ( optind == argc ){
		fprintf(stderr, "%s: no destination\n", program_name);
		return 1;
	}

	/* send marker */
	for ( int i = optind; i < argc; i++ ){
		struct in_addr ip_addr;
		in_port_t ip_port = 0; /* silence gcc stupid warning about initialized value, it is never used uninitialized */
		char* tmp;

		/* parse address */
		switch ( mode ){
		case MODE_UDP:
		case MODE_TCP:
			strtok(argv[i], ":");
			tmp = strtok(NULL, "");
			if ( !tmp ){
				fprintf(stderr, "%s: bad address `%s'\n", program_name, argv[i]);
				return 1;
			}
			inet_aton(argv[i], &ip_addr);
			ip_port = htons(atoi(tmp));

			fprintf(stderr, "%s: Sending marker to %s:%d\n", program_name, inet_ntoa(ip_addr), ntohs(ip_port));
			break;
		default:
		  fprintf(stderr,"What mode are you trying? I can handle TCP and UDP, whats %d?.\n",mode);
		}

		/* send */
		switch ( mode ){
		case MODE_UDP:
			send_udp(ip_addr, ip_port);
			break;
		case MODE_TCP:
			send_tcp(ip_addr, ip_port);
			break;
		default:
		  fprintf(stderr,"Selected method is not implemented.\n");

		}
	}

	return 0;
}
